### PR TITLE
Promote sysctls e2e test to Conformance

### PR DESCRIPTION
**Pull request by wgahnagl**
/area conformance 

@kubernetes/sig-architecture-pr-reviews @kubernetes/sig-node-pr-reviews @kubernetes/cncf-conformance-wg

### CODE REVIEW

**Comment by pacoxu** [comment]
@wgahnagl  need rebase.

/kind feature
/sig node
/ok-to-test
/priority important-soon
/triage accepted

**Comment by dims** [comment]
Looks like this was just recently merged : https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=NodeFeature:Sysctls

**Comment by dims** [comment]
Also please drop the `[NodeFeature:Sysctls]` tag

**Comment by dims** [comment]
Also you may want to first add this to `NodeConformance`

cc @SergeyKanzhelev @ehashman

**Comment by SergeyKanzhelev** [comment]
> Looks like this was just recently merged : https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=NodeFeature:Sysctls

Conformance needs [two weeks](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#user-content-conformance-test-requirements:~:text=it%20is%20stable%20and%20runs%20consistently%20(e.g.%2C%20no%20flakes)%2C%20and%20has%20been%20running%20for%20at%20least%20two%20weeks):

> it is stable and runs consistently (e.g., no flakes), and has been running for at least two weeks 

> Also please drop the `[NodeFeature:Sysctls]` tag

Without `NodeFeature`, we need NodeConformance so it will run in some of the dashboards

**Comment by dims** [comment]
@SergeyKanzhelev +1 to let it run for 2 weeks with `NodeConformance` before we consider it!

**Comment by spiffxp** [comment]
Is this expected to be supported across all clusters? Do nodes need to be configured in a certain way (beyond running Linux) to support this?

**Comment by pacoxu** [comment]
https://github.com/kubernetes/enhancements/issues/34#issuecomment-791888930
@ehashman  does this PR should be merged before code freeze？

**Comment by spiffxp** [comment]
/milestone v1.21
This is not subject to code freeze, but test freeze. 

I still haven't had time to answer my own questions above, if someone else does I would appreciate it.

**Comment by ehashman** [comment]
> Is this expected to be supported across all clusters? Do nodes need to be configured in a certain way (beyond running Linux) to support this?

I think this should be supported on all Linux machines, although it requires configuration in the kubelet https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/sysctl.md#abstract

@derekwaynecarr can you confirm if sysctls supported across all clusters with Linux nodes?

**Comment by spiffxp** [comment]
Is the kubelet configured this way by default in v1.21 or will cluster operators have to explicitly configure to support this?

**Comment by liggitt** [comment]
> Is the kubelet configured this way by default in v1.21 or will cluster operators have to explicitly configure to support this?

that's a good question. also relevant is whether this new test is expected to pass on a 1.20 and 1.19 kubelet or if it needs to be tagged in a way the skew tests can filter out (xref https://github.com/kubernetes/kubernetes/issues/99909#issuecomment-792347573)

**Comment by pacoxu** [comment]
kubelet feature-gate `Sysctls` is beta since 1.11. 
If there are nodes 1.19 or 1.20 with `--feature-gates Sysctls=false`, it would be a problem.

**Comment by liggitt** [comment]
> kubelet feature-gate `Sysctls` is beta since 1.11.
> If there are nodes 1.19 or 1.20 with `--feature-gates Sysctls=false`, it would be a problem.

Good point. This seems like a good place for `[MinimumKubeletVersion:1.21]` 

xref https://github.com/kubernetes/kubernetes/pull/100050 and https://github.com/kubernetes/community/pull/5622/files

File: test/e2e/common/node/sysctl.go
Comment (ehashman) [inline test/e2e/common/node/sysctl.go:74]: Per @liggitt can you please add the `[MinimumKubeletVersion:1.21]` tag to each of these descriptions?

```suggestion
	framework.ConformanceIt("should support sysctls [MinimumKubeletVersion:1.21]", func() {
```

**Comment by ehashman** [comment]
Conformance YAML verify job is failing, can you update?

File: test/conformance/testdata/conformance.yaml
Comment (pacoxu) [inline test/conformance/testdata/conformance.yaml:2110]: [MinimumKubeletVersion:1.21] add here as well.

The ci will check test/conformance/testdata/conformance.yaml as well.

**Comment by spiffxp** [comment]
/assign @derekwaynecarr @SergeyKanzhelev @ehashman 
I would like confirmation that this isn't the same situation as https://github.com/kubernetes/kubernetes/pull/97619#issuecomment-794518811 and https://github.com/kubernetes/kubernetes/pull/97619#issuecomment-794525453, where we had to revert.  Is this supported by default with no extra configuration if nodes are upgraded to or created at v1.21?

/assign @hh @johnbelamaric 
If we do proceed with this, I want to double check we're in agreement on the use of the tag `[MinimumKubeletVersion:1.21]` with respect to submissions to cncf/k8s-conformance contributors.  I would expect new v1.21 offerings to be certified without use of this tag.

Folks who use conformance tests on live clusters for some other reason (smoke test, health check, upgrade signal, etc.) may need to use this tag if their cluster has mixed node versions.

**Review by ehashman — COMMENTED** [review]
> I would like confirmation that this isn't the same situation as [#97619 (comment)](https://github.com/kubernetes/kubernetes/pull/97619#issuecomment-794518811) and [#97619 (comment)](https://github.com/kubernetes/kubernetes/pull/97619#issuecomment-794525453), where we had to revert. Is this supported by default with no extra configuration if nodes are upgraded to or created at v1.21?

Hi @spiffxp, yes this should be. These tests have been running in CI since this went beta in 1.11. We are cleaning things up this release and finally promoting to GA. The only code changes in this release were a removal of the feature flag.

/lgtm

**Comment by palnabarun** [comment]
Friendly reminder -- Tomorrow, March 24th Wednesday is v1.21 Test Freeze. :)

File: test/e2e/common/node/sysctl.go
Comment (BenTheElder) [inline test/e2e/common/node/sysctl.go:116]: nit: s/whitelisted/allow-listed/ or s/whitelisted/allowed/ in these test names
 
 The flag seems to be `allowed`, no reason to use `whitelisted` instead of `allowed` in the test names.
https://kubernetes.io/docs/tasks/administer-cluster/sysctl-cluster/#enabling-unsafe-sysctls
https://github.com/kubernetes/community/pull/5341

Comment (spiffxp) [reply test/e2e/common/node/sysctl.go:116]: I agree with this.  I won't block on this though, under the assumption that we can correct this post-test-freeze as a comment-only change (though it will result in changing the contents of conformance.yaml)

**Comment by spiffxp** [comment]
@ehashman 
>The only code changes in this release were a removal of the feature flag.

OK thanks that clarifies things for me, from that perspective the boxes I care about have been checked.

However, pulling out of https://kubernetes.slack.com/archives/C0BP8PW9G/p1616539528066300
>@derekwaynecarr  1 hour ago
>I will look in morning.  Unsafe sysctls should never be in conformance
>I need to look at what each test is doing

I'll wait for this review before putting on /approve

File: test/conformance/testdata/conformance.yaml
Comment (derekwaynecarr) [inline test/conformance/testdata/conformance.yaml:2094]: just recording notes as part of review.

this test checks for the following sysctl: "kernel.msgmax"

if we add this test to conformance, and this is later supported as a safe systctl by being namespaced, what is the policy for changing the enumerated sysctl in future?

runc see:
https://github.com/opencontainers/runc/blob/09523b79d01d032d2faeeb7eef4ad5a31eda528f/libcontainer/configs/validate/validator.go#L138

My memory is not as strong in this area as it should be but i am biased to not promote tests related to greylisting sysctls at this time, unless there is a particular policy on above question we accounted.

File: test/conformance/testdata/conformance.yaml
Comment (derekwaynecarr) [inline test/conformance/testdata/conformance.yaml:2103]: this test seems safe to promote.

File: test/conformance/testdata/conformance.yaml
Comment (derekwaynecarr) [reply test/conformance/testdata/conformance.yaml:2111]: i agree this test should be safe to promote.

File: test/conformance/testdata/conformance.yaml
Comment (derekwaynecarr) [inline test/conformance/testdata/conformance.yaml:2119]: this test appears to test the same functionality as the prior test since its using the same sysctl "kernel.shm_rmid_forced", is this an oversight?

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
I prefer we avoid elevating greylist sysctl tests into conformance until its understood what we should do if and when particular sysctls become namespaced.  Apologies if I am missing context as I was trying to refresh my own mental cache on sysctl behaviors while reviewing.

File: test/e2e/common/node/sysctl.go
Comment (wgahnagl) [reply test/e2e/common/node/sysctl.go]: fixed !

**Comment by spiffxp** [comment]
I am willing to approve if this PR is updated to not promote the two tests @derekwaynecarr has issues with. It is unclear to me whether this is something we could ask for an exception on should this not happen today. I will defer to @palnabarun for that decision

File: test/conformance/testdata/conformance.yaml
Comment (pacoxu) [reply test/conformance/testdata/conformance.yaml:2119]: `kernel.shm_rmid_forced` is in the safe set.

The current two test cases code are totally the same for `Sysctl, test sysctls` and `Sysctl, allow specified unsafe sysctls`.

**Comment by wgahnagl** [comment]
@derekwaynecarr does this look good?

**Comment by palnabarun** [comment]
Considering we are past code freeze, how critical is this promotion for v1.21? If we don't promote the tests, will it block releasing v1.21?

**Comment by spiffxp** [comment]
I don't have the KEP in front of me but AFAIK this is related to sysctl support going GA. It would be worse to promote to GA without conformance tests of any kind. We don't have automated enforcement of this policy.

It would be probably hard or painful to revert going to GA. I would lean towards having this land given the tests have had good CI signal

File: test/conformance/testdata/conformance.yaml
Comment (spiffxp) [reply test/conformance/testdata/conformance.yaml:2110]: This makes me question whether this test should be promoted then. No loss of coverage if we don't promote this?

**Comment by palnabarun** [comment]
@spiffxp -- that makes sense. As this relates to a GA graduating feature, we should try to include it.

**Comment by ehashman** [comment]
/milestone v1.21

then :)

**Comment by palnabarun** [comment]
@ehashman -- it was already in the milestone. :P

**Comment by ehashman** [comment]
SHHH YOU SAW NOTHING

**Comment by palnabarun** [comment]
Hahaha!! ;)

**Comment by derekwaynecarr** [comment]
Thank you for removing greylist tests.

/approve
/lgtm

**Comment by palnabarun** [comment]
/assign @johnbelamaric @smarterclayton 

(from conformance behavior approvers)

File: test/conformance/testdata/conformance.yaml
Comment (smarterclayton) [reply test/conformance/testdata/conformance.yaml:2110]: Holding approval until I see an answer here

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
Thanks for the attention all, appreciate that we can get some soak time with these tests over the weekend

**Pull request merged**
